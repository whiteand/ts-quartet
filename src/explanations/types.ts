import { KeyType, TPrimitiveSchema } from "../types";

export enum ExplanationSchemaType {
  And = "And",
  Any = "Any",
  Array = "Array",
  ArrayOf = "ArrayOf",
  Boolean = "Boolean",
  Finite = "Finite",
  Function = "Function",
  Max = "Max",
  MaxLength = "MaxLength",
  Min = "Min",
  MinLength = "MinLength",
  Negative = "Negative",
  Never = "Never",
  Not = "Not",
  NotANumber = "NotANumber",
  Number = "Number",
  Object = "Object",
  Pair = "Pair",
  Positive = "Positive",
  SafeInteger = "SafeInteger",
  String = "String",
  Symbol = "Symbol",
  Test = "Test",
  Variant = "Variant",
  Custom = "Custom"
}

export interface IAndExplanationSchema {
  type: ExplanationSchemaType.And;
  schemas: TExplanationSchema[];
}
export interface IAnyExplanationSchema {
  type: ExplanationSchemaType.Any;
}

export interface IArrayExplanationSchema {
  type: ExplanationSchemaType.Array;
}

export interface IArrayOfExplanationSchema {
  type: ExplanationSchemaType.ArrayOf;
  elementSchema: TExplanationSchema;
}

export interface IBooleanExplanationSchema {
  type: ExplanationSchemaType.Boolean;
}

export interface IFiniteExplanationSchema {
  type: ExplanationSchemaType.Finite;
}

export interface IFunctionExplanationSchema {
  type: ExplanationSchemaType.Function;
}
export interface IMaxExplanationSchema {
  type: ExplanationSchemaType.Max;
  maxValue: number;
  isExclusive: boolean;
}

export interface IMaxLengthExplanationSchema {
  type: ExplanationSchemaType.MaxLength;
  maxLength: number;
  isExclusive: boolean;
}
export interface IMinExplanationSchema {
  type: ExplanationSchemaType.Min;
  minValue: number;
  isExclusive: boolean;
}

export interface IMinLengthExplanationSchema {
  type: ExplanationSchemaType.MinLength;
  minLength: number;
  isExclusive: boolean;
}

export interface INegativeExplanationSchema {
  type: ExplanationSchemaType.Negative;
}
export interface INeverExplanationSchema {
  type: ExplanationSchemaType.Never;
}
export interface INotExplanationSchema {
  type: ExplanationSchemaType.Not;
  schema: TExplanationSchema;
}

export interface INumberExplanationSchema {
  type: ExplanationSchemaType.Number;
}

export interface INotANumberExplanationSchema {
  type: ExplanationSchemaType.NotANumber;
}

export interface IObjectExplanationSchemaWithoutRest {
  type: ExplanationSchemaType.Object;
  propsSchemas: {
    [key: string]: TExplanationSchema;
  };
}

export interface IObjectExplanationSchemaWithRest {
  type: ExplanationSchemaType.Object;
  propsSchemas: {
    [key: string]: TExplanationSchema;
  };
  "[v.rest]": TExplanationSchema;
  "[v.restOmit]"?: KeyType[];
}

export type TObjectExplanationSchema =
  | IObjectExplanationSchemaWithoutRest
  | IObjectExplanationSchemaWithRest;

export interface IPairExplanationSchema {
  type: ExplanationSchemaType.Pair;
  keyValueSchema: TExplanationSchema;
}

export interface IPositiveExplanationSchema {
  type: ExplanationSchemaType.Positive;
}

export interface ISafeIntegerExplanationSchema {
  type: ExplanationSchemaType.SafeInteger;
}

export interface IStringExplanationSchema {
  type: ExplanationSchemaType.String;
}

export interface ISymbolExplanationSchema {
  type: ExplanationSchemaType.Symbol;
}

export interface ITestExplanationSchema {
  type: ExplanationSchemaType.Test;
  description: string;
}

export interface IVariantExplanationSchema {
  type: ExplanationSchemaType.Variant;
  variants: TExplanationSchema[];
}

export interface ICustomExplanationSchema {
  type: ExplanationSchemaType.Custom;
  description: string;
  innerExplanations: any[];
}

export type TExplanationSchema =
  | TPrimitiveSchema
  | IAndExplanationSchema
  | IAnyExplanationSchema
  | IArrayExplanationSchema
  | IArrayOfExplanationSchema
  | IBooleanExplanationSchema
  | IFiniteExplanationSchema
  | IFunctionExplanationSchema
  | IMaxExplanationSchema
  | IMaxLengthExplanationSchema
  | IMinExplanationSchema
  | IMinLengthExplanationSchema
  | INegativeExplanationSchema
  | INeverExplanationSchema
  | INotExplanationSchema
  | INumberExplanationSchema
  | INotANumberExplanationSchema
  | TObjectExplanationSchema
  | IPairExplanationSchema
  | IPositiveExplanationSchema
  | ISafeIntegerExplanationSchema
  | IStringExplanationSchema
  | ISymbolExplanationSchema
  | ITestExplanationSchema
  | IVariantExplanationSchema
  | ICustomExplanationSchema;

export interface IExplanation {
  value: any;
  path: KeyType[];
  schema: TExplanationSchema;
  innerExplanations: IExplanation[];
}
