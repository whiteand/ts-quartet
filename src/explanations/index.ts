export * from "./types";
export * from "./schemaToExplanationSchema";
