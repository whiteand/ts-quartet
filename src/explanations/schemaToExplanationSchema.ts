/* tslint:disable:object-literal-sort-keys */
import { SchemaType } from "../schemas";
import { KeyType, TPrimitiveSchema, TSchema } from "../types";
import { ExplanationSchemaType, TExplanationSchema } from "./types";

const SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE = {
  [SchemaType.And]: ExplanationSchemaType.And as ExplanationSchemaType.And,
  [SchemaType.Any]: ExplanationSchemaType.Any as ExplanationSchemaType.Any,
  [SchemaType.Array]: ExplanationSchemaType.Array as ExplanationSchemaType.Array,
  [SchemaType.ArrayOf]: ExplanationSchemaType.ArrayOf as ExplanationSchemaType.ArrayOf,
  [SchemaType.Boolean]: ExplanationSchemaType.Boolean as ExplanationSchemaType.Boolean,
  [SchemaType.Finite]: ExplanationSchemaType.Finite as ExplanationSchemaType.Finite,
  [SchemaType.Function]: ExplanationSchemaType.Function as ExplanationSchemaType.Function,
  [SchemaType.Max]: ExplanationSchemaType.Max as ExplanationSchemaType.Max,
  [SchemaType.MaxLength]: ExplanationSchemaType.MaxLength as ExplanationSchemaType.MaxLength,
  [SchemaType.Min]: ExplanationSchemaType.Min as ExplanationSchemaType.Min,
  [SchemaType.MinLength]: ExplanationSchemaType.MinLength as ExplanationSchemaType.MinLength,
  [SchemaType.Negative]: ExplanationSchemaType.Negative as ExplanationSchemaType.Negative,
  [SchemaType.Never]: ExplanationSchemaType.Never as ExplanationSchemaType.Never,
  [SchemaType.Not]: ExplanationSchemaType.Not as ExplanationSchemaType.Not,
  [SchemaType.NotANumber]: ExplanationSchemaType.NotANumber as ExplanationSchemaType.NotANumber,
  [SchemaType.Number]: ExplanationSchemaType.Number as ExplanationSchemaType.Number,
  [SchemaType.Object]: ExplanationSchemaType.Object as ExplanationSchemaType.Object,
  [SchemaType.Pair]: ExplanationSchemaType.Pair as ExplanationSchemaType.Pair,
  [SchemaType.Positive]: ExplanationSchemaType.Positive as ExplanationSchemaType.Positive,
  [SchemaType.SafeInteger]: ExplanationSchemaType.SafeInteger as ExplanationSchemaType.SafeInteger,
  [SchemaType.String]: ExplanationSchemaType.String as ExplanationSchemaType.String,
  [SchemaType.Symbol]: ExplanationSchemaType.Symbol as ExplanationSchemaType.Symbol,
  [SchemaType.Test]: ExplanationSchemaType.Test as ExplanationSchemaType.Test,
  [SchemaType.Variant]: ExplanationSchemaType.Variant as ExplanationSchemaType.Variant,
  [SchemaType.Custom]: ExplanationSchemaType.Custom as ExplanationSchemaType.Custom
};

export function schemaToExplanationSchema(schema: TSchema): TExplanationSchema {
  if (typeof schema !== "object" || schema === null) {
    return {
      type: ExplanationSchemaType.Primitive,
      value: schema as TPrimitiveSchema
    };
  }
  switch (schema.type) {
    case SchemaType.And:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.And],
        schemas: schema.schemas.map(schemaToExplanationSchema)
      };
    case SchemaType.Any:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Any]
      };
    case SchemaType.Array:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Array]
      };
    case SchemaType.ArrayOf:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.ArrayOf],
        elementSchema: schemaToExplanationSchema(schema.elementSchema)
      };
    case SchemaType.Boolean:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Boolean]
      };
    case SchemaType.Finite:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Finite]
      };
    case SchemaType.Function:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Function]
      };
    case SchemaType.Max:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Max],
        maxValue: schema.maxValue,
        isExclusive: schema.isExclusive
      };
    case SchemaType.MaxLength:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.MaxLength],
        maxLength: schema.maxLength,
        isExclusive: schema.isExclusive
      };
    case SchemaType.Min:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Min],
        minValue: schema.minValue,
        isExclusive: schema.isExclusive
      };
    case SchemaType.MinLength:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.MinLength],
        minLength: schema.minLength,
        isExclusive: schema.isExclusive
      };
    case SchemaType.Negative:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Negative]
      };
    case SchemaType.Never:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Never]
      };
    case SchemaType.Not:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Not],
        schema: schemaToExplanationSchema(schema.schema)
      };
    case SchemaType.NotANumber:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.NotANumber]
      };
    case SchemaType.Number:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Number]
      };
    case SchemaType.Object:
      const propsSchemas: Record<KeyType, TExplanationSchema> = Object.create(
        null
      );
      const { props } = schema;
      for (let i = 0; i < props.length; i++) {
        const prop = props[i];
        propsSchemas[prop] = schemaToExplanationSchema(
          schema.propsSchemas[prop as string]
        );
      }
      if (!schema.hasRestValidator) {
        return {
          type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Object],
          propsSchemas
        };
      }
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Object],
        propsSchemas,
        "[v.rest]": schemaToExplanationSchema(schema.rest),
        "[v.restOmit]": Object.keys(schema.restOmitDict)
      };
    case SchemaType.Pair:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Pair],
        keyValueSchema: schemaToExplanationSchema(schema.keyValueSchema)
      };
    case SchemaType.Positive:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Positive]
      };
    case SchemaType.SafeInteger:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.SafeInteger]
      };
    case SchemaType.String:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.String]
      };
    case SchemaType.Symbol:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Symbol]
      };
    case SchemaType.Test:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Test],
        tester: schema.tester
      };
    case SchemaType.Variant:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Variant],
        variants: schema.variants.map(schemaToExplanationSchema)
      };
    case SchemaType.Custom:
      return {
        type: SCHEMA_TYPE_TO_EXPLANATION_SCHEMA_TYPE[SchemaType.Custom],
        customValidator: schema.customValidator
      };
  }
}
