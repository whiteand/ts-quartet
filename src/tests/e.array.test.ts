/* tslint:disable:object-literal-sort-keys */
import { e as v } from "..";
import { ExplanationSchemaType } from "../explanations";
import { testValidatorWithExplanations } from "./testValidatorWithExplanations";

describe("v.array", () => {
  test("v.array", () => {
    testValidatorWithExplanations(
      v(v.array),
      [[], [1, 2, "3"]],
      [
        [
          {},
          [
            {
              value: {},
              path: [],
              schema: {
                type: ExplanationSchemaType.Array
              }
            }
          ]
        ],
        [
          { length: 10 },
          [
            {
              value: { length: 10 },
              path: [],
              schema: {
                type: ExplanationSchemaType.Array
              }
            }
          ]
        ],
        [
          "Andrew",
          [
            {
              value: "Andrew",
              path: [],
              schema: {
                type: ExplanationSchemaType.Array
              }
            }
          ]
        ]
      ]
    );
  });
});
