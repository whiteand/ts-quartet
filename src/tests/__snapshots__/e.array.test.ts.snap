// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v.array v.array 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "type": "Array",
    },
    "value": Object {},
  },
]
`;

exports[`v.array v.array 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "type": "Array",
    },
    "value": Object {
      "length": 10,
    },
  },
]
`;

exports[`v.array v.array 3`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "type": "Array",
    },
    "value": "Andrew",
  },
]
`;
